### PR TITLE
004 view tasks (all tasks and archived)

### PR DESCRIPTION
**Pull request by noxiousghost**
This PR is about adding the feature for viewing tasks from the list. Here are the changes made:
- In index.ts file, i have created a command called show that accepts two options (-a or --all and -ar and --archived).
- In tasks.controller.ts file, I created a method called viewTask which include logic for viewing all tasks. 

Summary:
- [x] User can view all task that are not archived by using ```add -a``` or ```add --all``` command.
- [x] User can view all task that are  archived by using ```add -ar``` or ```add --archived``` command.

### CODE REVIEW

File: src/controllers/tasks.controller.ts [hunk 1]
```diff
@@ -18,8 +19,8 @@ export const createTask = async (options: Omit<Task, 'id' | 'status' | 'isArchiv
       console.error('Validation failed:', parsedTask.error.flatten());
       return;
     }
-    const readJsonRes = await readJsonFile();
-    tasks = readJsonRes;
+    tasks = await readJsonFile();
```
Comment (arjandhakal) [inline src/controllers/tasks.controller.ts]: lets define and declare task here, `const tasks = await readJsonFile()`

File: src/controllers/tasks.controller.ts [hunk 2]
```diff
@@ -29,3 +30,34 @@ export const createTask = async (options: Omit<Task, 'id' | 'status' | 'isArchiv
...
+    // checking if task is archived or not
+    if (options.archived) {
+      tasks = tasks.filter((task) => task.isArchived);
+    } else if (options.all) {
+      tasks = tasks.filter((task) => !task.isArchived);
+    }
+
+    // displaying all the tasks result
+    if (tasks.length > 0) {
```
Comment (arjandhakal) [inline src/controllers/tasks.controller.ts]: Better to have a separate function, just for displaying/printing task

File: src/controllers/tasks.controller.ts [hunk 3]
```diff
@@ -29,3 +30,34 @@ export const createTask = async (options: Omit<Task, 'id' | 'status' | 'isArchiv
...
   }
 };
+
+export const viewTasks = async (options: { all: string; archived: string }): Promise<void> => {
+  try {
+    if (Object.keys(options).length === 0) {
+      throw error('No options provided');
+    }
+    let tasks = [];
+    tasks = await readJsonFile();
```
Comment (arjandhakal) [inline src/controllers/tasks.controller.ts]: lets keep task as const

File: src/controllers/tasks.controller.ts [hunk 4]
```diff
@@ -29,3 +30,34 @@ export const createTask = async (options: Omit<Task, 'id' | 'status' | 'isArchiv
...
+    // checking if task is archived or not
+    if (options.archived) {
+      tasks = tasks.filter((task) => task.isArchived);
```
Comment (arjandhakal) [inline src/controllers/tasks.controller.ts]: I don't think we should mutate task here, it's always good to try and minimize state mutation. It makes it easy to read the code later on and makes it easy to keep track of what value each variable holds. 

An alternative approach might be:
```
if(options.archived){
displayTasks(tasks.filter(t => t.isArchived))
}
```

Comment (arjandhakal) [reply src/controllers/tasks.controller.ts]: This way we don't need to track so hard about how tasks variable has changed. This is a simple example, but you can see how if we have multiple variables such as this making it diffcult to track

**Comment by arjandhakal** [comment]
How would I be able to display both archived and non-archived ? Current approach does not make it possible

**Review by arjandhakal — APPROVED** [review]

File: src/utils/displayTasks.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { Task } from '@src/model/zod.schema';
+export const displayTasks = (tasks: Task[]) => {
+  if (tasks.length === 0) {
```
Comment (Nijarshrestha) [inline src/utils/displayTasks.ts]: We can use the shorthand method for this, `!tasks.length`

File: src/controllers/tasks.controller.ts [hunk 6]
```diff
@@ -29,3 +28,21 @@ export const createTask = async (options: Omit<Task, 'id' | 'status' | 'isArchiv
     console.error(error);
   }
 };
+
+export const viewTasks = async (options: { all: string; archived: string }): Promise<void> => {
+  try {
+    const tasks = await readJsonFile();
+    if (options.archived) {
```
Comment (Nijarshrestha) [inline src/controllers/tasks.controller.ts:36]:
```
 try {
    const tasks = await readJsonFile();
    if (options.archived) {
      // only archived tasks
      displayTasks(tasks.filter((task) => task.isArchived));
    } else if (options.all) {
      // all tasks
      displayTasks(tasks);
    } else {
      // default--> show non archived tasks
      displayTasks(tasks.filter((task) => !task.isArchived));
    }
```

Can this be written in `if early return` way
```
 try {
    const tasks = await readJsonFile();
    if (options.archived) {
      // only archived tasks
      return displayTasks(tasks.filter((task) => task.isArchived)); 
    }  if (options.all) {
      // all tasks
      return displayTasks(tasks);
    } 
      // default--> show non archived tasks
      return displayTasks(tasks.filter((task) => !task.isArchived));
    }
```

In my opinion, it's much cleaner and easier to add more if condition.

File: src/controllers/tasks.controller.ts [hunk 7]
```diff
@@ -29,3 +28,21 @@ export const createTask = async (options: Omit<Task, 'id' | 'status' | 'isArchiv
...
+      // only archived tasks
+      displayTasks(tasks.filter((task) => task.isArchived));
+    } else if (options.all) {
```
Comment (Nijarshrestha) [inline src/controllers/tasks.controller.ts]: is `options.all` boolean or string value ?

Comment (noxiousghost) [reply src/controllers/tasks.controller.ts]: options is an object; if --all or -a flag is passed while running the application then options.all becomes boolean with truth value. same goes to options.archived.

**Review by Nijarshrestha — APPROVED** [review]

**Pull request merged**